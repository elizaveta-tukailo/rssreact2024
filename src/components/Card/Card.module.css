.people__item {
  padding: 16px;
  border-radius: 8px;
  border: 1px solid #c8c8c8;
}
