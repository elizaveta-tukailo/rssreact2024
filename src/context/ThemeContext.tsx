import {
  createContext,
  ReactNode,
  useContext,
  useMemo,
  useState,
  useCallback,
} from 'react';
type Theme = 'light' | 'dark';
interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}
export const ThemeContext = createContext<ThemeContextType | undefined>(
  undefined
);
export function useTheme() {
  const context = useContext(ThemeContext);
  if (context === undefined) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
}
interface ThemeProviderProps {
  children: ReactNode;
}
export function ThemeProvider({ children }: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>('light');
  const toggleTheme = useCallback(() => {
    setTheme((prevTheme) => (prevTheme === 'light' ? 'dark' : 'light'));
  }, []);
  const value = useMemo(() => ({ theme, toggleTheme }), [theme, toggleTheme]);
  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
}
