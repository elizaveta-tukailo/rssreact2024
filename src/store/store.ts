import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import { characterApi } from './../services/character';

export const store = configureStore({
  reducer: {
    [characterApi.reducerPath]: characterApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(characterApi.middleware),
});

setupListeners(store.dispatch);
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
