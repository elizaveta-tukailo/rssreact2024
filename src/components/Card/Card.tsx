import styles from './card.module.css';
import Loader from '../Loader';
import { useDispatch } from 'react-redux';
import {
  setCharacterData,
  setIsClosed,
} from '../../store/reducers/characterDetailSlice';
import { useTheme } from '../../context/ThemeContext';
import { useGetCharacterQuery } from '../../services/character';
import { useEffect } from 'react';
import { setPageCharacters } from '../../store/reducers/pageCharactersSlice';
import { useRouter } from 'next/router';

const Card = () => {
  const { theme } = useTheme();
  const dispatch = useDispatch();
  const router = useRouter();
  const { details } = router.query;
  const {
    data: character,
    error,
    isFetching,
  } = useGetCharacterQuery(Number(details));

  useEffect(() => {
    if (character) {
      dispatch(setPageCharacters(character));
      dispatch(setIsClosed({ isClosed: false, characterId: character.id }));
    }
  }, [dispatch, character]);

  const clickCloseModal = () => {
    dispatch(setIsClosed({ isClosed: true, characterId: 0 }));
    dispatch(setCharacterData([]));
  };

  if (!character) {
    return <div>No data found!</div>;
  }
  return (
    <>
      {isFetching ? (
        <div data-testid="loader">
          <Loader />
        </div>
      ) : null}
      {!error ? (
        <>
          <div
            className={`${styles['cardItemWrap']} ${styles[theme]}`}
            data-testid="detail-card"
          >
            <div className={styles['cardItem']}>
              <div className={styles['cardItemId']}>
                Card ID: {character.id}
              </div>
              <h3 className={styles['cardItemTitle']}> {character.name} </h3>
              <button
                className={styles['cardItemBtn']}
                onClick={clickCloseModal}
              ></button>
            </div>
            <div className={styles['cardItemRow']}>
              <div className={styles['cardItemImage']}>
                <img src={character.image}></img>
              </div>
              <div className={styles['cardItemInfo']}>
                <div className={styles['cardItemInfoItem']}>
                  <span>Species:</span> {character.species}
                </div>
                <div className={styles['cardItemInfoItem']}>
                  <span>Status:</span> {character.status}
                </div>
                <div className={styles['cardItemInfoItem']}>
                  <span> Gender:</span> {character.gender}
                </div>
                <div className={styles['cardItemInfoItem']}>
                  <span>Planet:</span> {character.location.name}
                </div>
              </div>
            </div>
          </div>
        </>
      ) : (
        <div className={styles.error}>Error</div>
      )}
    </>
  );
};
export default Card;
