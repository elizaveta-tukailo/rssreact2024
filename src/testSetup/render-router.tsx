import { store } from '../store/store';
import { render } from '@testing-library/react';
import { ReactNode } from 'react';

import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { ThemeProvider } from '../context/ThemeContext';

export const renderWithRouter = (
  ui: ReactNode,
  { route = '/', path = '/' } = {}
) => {
  window.history.pushState({}, 'Test page', route);

  return render(
    <Provider store={store}>
      <ThemeProvider>
        <MemoryRouter initialEntries={[route]}>
          <Routes>
            <Route path={path} element={ui} />
          </Routes>
        </MemoryRouter>
      </ThemeProvider>
    </Provider>
  );
};
