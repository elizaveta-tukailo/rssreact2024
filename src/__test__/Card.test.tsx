import '@testing-library/jest-dom';
const { expect, describe, it } = require('@jest/globals');
import { MemoryRouter } from 'react-router-dom';
import Card from '../components/Card';
import { ThemeProvider } from '../context/ThemeContext';
import { Provider } from 'react-redux';
import { store } from '../store/store';
import { character } from './__mocks__/charactersMockData';
import { render, screen } from '@testing-library/react';

import * as reduxHooks from 'react-redux';
import { renderWithRouter } from '../testSetup/render-router';

jest.mock('react-redux');

describe('Testing card component', () => {
  it('renders loading state', () => {
    renderWithRouter(<Card />, {
      route: '/details/1',
    });

    expect(screen.getByTestId('loader')).toBeInTheDocument();
  });
  it('Match snapshot', () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue(character);

    const component = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/details/1`]}>
          <ThemeProvider>
            <Card />
          </ThemeProvider>
        </MemoryRouter>
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });
  it('Loader should appear while data is fetching', async () => {
    renderWithRouter(<Card />, {
      route: '/details/1',
    });

    const loader = await screen.findByTestId('loader');
    expect(loader).toBeInTheDocument();
  });
  it('Check if the detailed card component correctly displays the character data;', async () => {
    renderWithRouter(<Card />, {
      route: '/details/1',
    });

    const characterName = await screen.getByText(/Morty/i);
    expect(characterName).toBeInTheDocument();
  });
});
