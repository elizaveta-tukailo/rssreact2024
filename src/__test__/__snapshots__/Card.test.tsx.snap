// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component Card component renders correctly 1`] = `
<div>
  No data found!
</div>
`;
